### ISSUE TITLE
Jenkins tests with the address sanitizer are failing 

### ISSUE THREAD

**Issue by poojanilangekar** [cmu-db/peloton#1154]
Even though Jenkins says all test have passed, we have a total of 59 tests failing due to memory leaks. This is something @pervazea noticed on my PR. But I just checked that all the other PR and the master repo is also leaking the same memory. The [master](http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/master/5/pipeline/18) also leaks memory. Strangely, I wasn't able to reproduce it on my vagrant machine. Pervaze was unable to reproduce this with the same configuration. 

This happens on `Build / Ubuntu Xenial/gcc-5.4.0/llvm-3.7.1 (Debug/Test)` for the `cd build && make check -j4 || true` test.

**Comment by pervazea** [comment]
I'll talk to Chad tomorrow. We should be paying attention to the return value from the tests. Jenkins seems to have no mechanism to run a set of things, then report whether one of N items failed. You either stop as soon as one invocation fails, or ignore the error code and continue running. We should be able to write our own wrapper to address this, in due course.


At the moment, we've been trying to resolve the issue that Jenkins stalls waiting for "executors", even though resources are idle. Until this is fixed, we can't scale things up properly.


________________________________

**Comment by pervazea** [comment]
>Also, I don't think the build command should be make check -j4 || true since that hides the errors.

Completely agree.


________________________________

**Comment by crd477** [comment]
Disregard what I said previously about ```make -k```; it was nonsensical.

The purpose of the ```||true``` is so that later steps, e.g., ```psql_test.sh``` and ```jdbc_validator.py``` can still run even if one of the unit tests fails.  I talked with @pervazea a bit about it this morning and I'm no longer even convinced that the current test invocation and collection is doing the right thing because the test results bar graph reports no failed tests even though the console logs clearly do.  It's on my short list to look into.

**Comment by pervazea** [comment]
On the status of PR1134

We see in the Jenkins builds:
[](url)http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/activity?branch=PR-1134
Failures:
1 - Xenial, Debug - Peloton failed with assertion failure
4 - same as 1.
6 - Trusty, Release - failed psql_test.sh, with data being inconsistent with reference output
12 - failed in communicating with Github. (Ignore this, nothing to do with Peloton)

I've tried to reproduce the failure(s) on my laptop, with no success, although I've not run the test a large number of times yet.

Using the docker container that Jenkins uses for these builds, it is possible to reproduce the assertion error. The problem is not 100% reproducible. Running the psql test in a loop, it usually fails after 4-6 iterations, but may not fail.

There are some additional complications to tracking this down:
1. No corefile is generated. I've checked the obvious settings with respect to corefiles and they are correct.

2. Running directly under GDB is no better. The message is approx:
During startup program terminated with signal SIGABRT, Killed.

Upon inspection, no process is running, so no backtrace or state examination is possible.

**Comment by poojanilangekar** [comment]
@pervazea I think we can add a breakpoint right before the assertion and check the value of `GetLastReaderCommitId` and then do a backtrace when it doesn't match the commit_id. Can we test this tomorrow?

**Comment by pervazea** [comment]
We can try this. I believe I did exactly that, but it did not work.
Now, I was not sure that when I first did this, which type of build I had, since gdb complained about unknown filename. So, when we do this again, we should:

- delete and recreate a debug build
- then try inserting the breakpoint and running

Chad also suggested looking at lldb if gdb won't cooperate.

**Comment by pervazea** [comment]
More on the psql failure.
1. Pooja thinks she has a fix, related to how the commit id is saved. 

1. The failure we've been seeing, is intermittent. We determined today, that it only manifests when the Jenkins machine has a higher load. Even then, not 100% of the time. 

1. Running via gdb in the Docker container, still doesn't work as expected. Somehow, the process terminates without gdb being able to stop it, so the assertion triggers, and there is no stack trace or core dump to examine. The last line, below, is the symptom of this problem.

peloton: /peloton/src/concurrency/timestamp_ordering_transaction_manager.cpp:641: virtual void peloton::concurrency::TimestampOrderingTransactionManager::PerformDelete(peloton::concurrency::TransactionContext*, const peloton::ItemPointer&, const peloton::ItemPointer&): Assertion `(GetLastReaderCommitId(tile_group_header, old_location.offset) == current_txn->GetCommitId())' failed.
During startup program terminated with signal SIGABRT, Aborted.

**Comment by poojanilangekar** [comment]
I tried adding a memory fence before the` SetLastReaderCommitId` function. However, this didn't fix the problem either. So now the assertion must necessarily pass. Not sure if there is anything else I am missing.

**Comment by pervazea** [comment]
I tried on Friday to recreate this problem without success. Have solicited Chad's help to recreate. 

I am not familiar with the interacting bodies of code, but have some speculation. Suggestions on what to instrument or how to debug would be welcome. 

I suspect that after the assertion fails, the transaction level garbage collector executes and is destroying / freeing the structures GDB is trying to access, in order to honour the breakpoint. Is there a way to turn off the garbage collector, so we can examine the system state?

**Comment by crd477** [comment]
A quick note on how to induce enough load to reproduce:
1. ssh to jenkins
1. in one window:
```docker run --rm -it pr1134:trusty```
then inside that container, do:
```cd /peloton/build; while : ; do make -j4 check; done```
1. (repeat step 2 in second window)
1. in a third window
```docker run -it pr1134:trusty```
then inside that container, do: 
```
cd /peloton/build
bash ../script/testing/psql/psql_test.sh
```

**Comment by pervazea** [comment]
We recreated this with a couple of additional log statements right before the assert that trips:
- last reader commit id was 0
- transaction commit id was a non-zero, reasonable value.
The assertion did not trip. (This is odd, implying that the values changed between the time the LOG_INFO statements read them and the assertion was executed).

The psql test failed, with the update statement being executed matched 2 records, rather than 1.

**Comment by pervazea** [comment]
2018-02-26 19:44:48 [src/concurrency/timestamp_ordering_transaction_manager.cpp:642:PerformDelete] INFO  - last rdr commit id 0
2018-02-26 19:44:48 [src/concurrency/timestamp_ordering_transaction_manager.cpp:645:PerformDelete] INFO  - cur commit id 1640677507276
peloton: /peloton/src/concurrency/timestamp_ordering_transaction_manager.cpp:648: virtual void peloton::concurrency::TimestampOrderingTransactionManager::PerformDelete(peloton::concurrency::TransactionContext*, const peloton::ItemPointer&, const peloton::ItemPointer&): Assertion `(GetLastReaderCommitId(tile_group_header, old_location.offset) == current_txn->GetCommitId())' failed.

**Comment by pervazea** [comment]
- The problem appears to be independent of codegen.
- No success when running under GDB (yet)

**Comment by pervazea** [comment]
Success with GDB!
When GDB started, it had an error about being unable to turn off address space randomization. This was significant. Docker containers prevent various operations by default. To make GDB work, invoke the docker container like so (where pr1134:trusty is the docker container I'm using):

docker run --security-opt seccomp=unconfined --rm -it pr1134:trusty

Then GDB can correctly set breakpoints, and we break on them successfully.

**Comment by pervazea** [comment]
The following approach looks to be working (but haven't examined carefully yet). Diff attached. Make Peloton think it has lots of CPUs, even though it doesn't. psql test failing on my laptop, whereas without this change, worked fine.

Thinking that a cleaned up version where we can manually pass in the value (when desired) via a command line parameter, might be useful for testing and flushing out problems of this type. 

diff --git a/src/common/init.cpp b/src/common/init.cpp
index 17e7606..bb0d2af 100644
--- a/src/common/init.cpp
+++ b/src/common/init.cpp
@@ -30,14 +30,15 @@ namespace peloton {
 ThreadPool thread_pool;
 
 void PelotonInit::Initialize() {
-  CONNECTION_THREAD_COUNT = std::thread::hardware_concurrency();
+  // CONNECTION_THREAD_COUNT = std::thread::hardware_concurrency();
+  CONNECTION_THREAD_COUNT = 24;
   LOGGING_THREAD_COUNT = 1;
   GC_THREAD_COUNT = 1;
   EPOCH_THREAD_COUNT = 1;
   MAX_CONCURRENCY = 10;
 
   // set max thread number.
-  thread_pool.Initialize(0, std::thread::hardware_concurrency() + 3);
+  thread_pool.Initialize(0, CONNECTION_THREAD_COUNT + 3);
 
   // start worker pool
   threadpool::MonoQueuePool::GetInstance().Startup();
@@ -47,7 +48,7 @@ void PelotonInit::Initialize() {
     threadpool::MonoQueuePool::GetBrainInstance().Startup();
   }
 
-  int parallelism = (std::thread::hardware_concurrency() + 3) / 4;
+  int parallelism = (CONNECTION_THREAD_COUNT + 3) / 4;
   storage::DataTable::SetActiveTileGroupCount(parallelism);
   storage::DataTable::SetActiveIndirectionArrayCount(parallelism);

**Comment by poojanilangekar** [comment]
I think we can do that by adding a parameter to the SettingsManager

**Comment by poojanilangekar** [comment]
@yingjunwu I think the bug lies in the `AbortTransaction` function of the `TimeStampOrderingTransactionManager`. When the psql_test.sh failed. I tried to find the version of the tuple it returned and the transaction id, begin cid and end cid of the old location. (Some background, the tuple is from the previous update statement which failed, so it contains the value <1,1000> which failed because the primary key constraint was violated).

Here are the contents of old_location:

Invalid Tuple <id = 1, year = 1000> 

Transaction ID 730144440378
Begin Commit ID 721554505772
End Commit ID 18446744073709551615

Current transaction ID 730144440378

I think we are having this issue because in the code block from 986 - 1018, we never modified the end CID of the UPDATE object. Ideally we should've set it to INVALID_CID to ensure that it is not visible to any transaction. What do you think?

**Comment by yingjunwu** [comment]
Do you mean that we did not set the end commit id of the newly created version (for the updated tuple) back to INVALID_CID?

**Comment by poojanilangekar** [comment]
Yes.

**Comment by poojanilangekar** [comment]
@yingjunwu I tried the fix I suggested. But that didn't work. I have a question regarding the way the `PerformUpdate` function is designed. Why is the index being updated even before the transaction commits? I specifically mean this line:

`    UNUSED_ATTRIBUTE auto res =
        AtomicUpdateItemPointer(index_entry_ptr, new_location);`

Could this be causing the concurrency bug? Can this update be made when the transaction commits?
